File: exercise-reference/completed/src/documentAI.ts
import { DocumentAi } from "@abbyy-sdk/document-ai";
import fs from 'fs';

// Initialize the Document AI client with the API key
const documentAi = new DocumentAi({
  apiKeyAuth: "abbyy_KaADPZtabT2IITvoTozOzZqsUMFFFPb4A63BiFDsDuX27vreL",
});

// Interface for utility bill data
export interface UtilityBillData {
  fullName: string;
  accountNumber: string;
  address: {
    street: string;
  };
  issueDate: string;
  utilityProvider: string;
}

/**
 * Process a utility bill image using Document AI
 * @param imageUrl URL of the utility bill image
 */
export async function processUtilityBill(imageUrl: string): Promise<UtilityBillData> {
  try {
    const extractRequest = await documentAi.models.utilityBill.beginFieldExtraction({
      inputSource: {
        base64EncodedContent: readFileAsBase64(imageUrl),
        name: "utility-bill.jpg"
      },
    });
    
    const documentId = extractRequest.documents?.[0]?.id;
    if (!documentId) {
      throw new Error("Failed to process document");
    }
    
    // Poll for the result
    let processed = false;
    let response;
    while (!processed) {
      await new Promise((resolve) => setTimeout(resolve, 500));
      response = await documentAi.models.utilityBill.getExtractedFields({
        documentId: documentId,
      });
      processed = response.utilityBill?.meta.status === "Processed";
    }

    // Extract relevant fields
    let responseFields = response?.utilityBill?.fields;
    return {
      fullName: responseFields?.customer?.name || "Unknown",
      accountNumber: responseFields?.customer?.accountNumber || "Unknown",
      address: {
        street: responseFields?.issuer?.address || "Unknown"
      },
      issueDate: responseFields?.billDate?.toString() || "",
      utilityProvider: responseFields?.issuer?.name || ""
    };
  } catch (error) {
    console.error("Error processing utility bill:", error);
    throw error;
  }
}

function readFileAsBase64(filePath: string): string {
  const file = fs.readFileSync(filePath);
  return file.toString("base64");
}

// Export the Document AI client for direct use if needed
export default DocumentAi; 

File: py/exercise-reference/completed/document_ai.py
import base64
import asyncio
from typing import Dict, Any
from dataclasses import dataclass

# Note: In a real implementation, you would import the ABBYY SDK
# from abbyy_sdk import DocumentAi

@dataclass
class UtilityBillData:
    full_name: str
    account_number: str
    address: Dict[str, str]
    issue_date: str
    utility_provider: str

# Mock Document AI client for demonstration
# In a real implementation, you would use the actual ABBYY SDK
class MockDocumentAi:
    def __init__(self, api_key: str):
        self.api_key = api_key
    
    class Models:
        class UtilityBill:
            async def begin_field_extraction(self, input_source: Dict[str, Any]):
                # Mock response
                return {
                    "documents": [{"id": "mock-document-id"}]
                }
            
            async def get_extracted_fields(self, document_id: str):
                # Mock processed response
                return {
                    "utilityBill": {
                        "meta": {"status": "Processed"},
                        "fields": {
                            "customer": {
                                "name": "John Doe",
                                "accountNumber": "1234567890"
                            },
                            "issuer": {
                                "address": "123 Main St",
                                "name": "Electricity Company"
                            },
                            "billDate": "2024-01-01"
                        }
                    }
                }
    
    @property
    def models(self):
        return self.Models()

# Initialize the Document AI client with the API key
# In production, use: document_ai = DocumentAi(api_key="your_api_key_here")
document_ai = MockDocumentAi(api_key="abbyy_KaADPZtabT2IITvoTozOzZqsUMFFFPb4A63BiFDsDuX27vreL")

def read_file_as_base64(file_path: str) -> str:
    """Read a file and return its base64 encoded content"""
    with open(file_path, 'rb') as file:
        file_content = file.read()
        return base64.b64encode(file_content).decode('utf-8')

async def process_utility_bill(image_path: str) -> UtilityBillData:
    """
    Process a utility bill image using Document AI
    Args:
        image_path: Path to the utility bill image file
    Returns:
        UtilityBillData object with extracted information
    """
    try:
        extract_request = await document_ai.models.utility_bill.begin_field_extraction({
            "inputSource": {
                "base64EncodedContent": read_file_as_base64(image_path),
                "name": "utility-bill.jpg"
            }
        })
        
        document_id = extract_request.get("documents", [{}])[0].get("id")
        if not document_id:
            raise Exception("Failed to process document")
        
        # Poll for the result
        processed = False
        response = None
        while not processed:
            await asyncio.sleep(0.5)  # Wait 500ms
            response = await document_ai.models.utility_bill.get_extracted_fields({
                "documentId": document_id
            })
            processed = response.get("utilityBill", {}).get("meta", {}).get("status") == "Processed"

        # Extract relevant fields
        response_fields = response.get("utilityBill", {}).get("fields", {})
        return UtilityBillData(
            full_name=response_fields.get("customer", {}).get("name", "Unknown"),
            account_number=response_fields.get("customer", {}).get("accountNumber", "Unknown"),
            address={
                "street": response_fields.get("issuer", {}).get("address", "Unknown")
            },
            issue_date=str(response_fields.get("billDate", "")),
            utility_provider=response_fields.get("issuer", {}).get("name", "")
        )
    except Exception as error:
        print(f"Error processing utility bill: {error}")
        raise error 
